src/Util/SmartPaginatorFactory.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\DoctrineORMAdminBundle\Util;
+
+use Doctrine\ORM\Tools\Pagination\CountWalker;
+use Doctrine\ORM\Tools\Pagination\Paginator;
+use Sonata\DoctrineORMAdminBundle\Datagrid\ProxyQuery;
+use Sonata\DoctrineORMAdminBundle\Datagrid\ProxyQueryInterface;
+
+/**
+ * @internal
+ */
+final class SmartPaginatorFactory
+{
+    /**
+     * NEXT_MAJOR: Replace ProxyQuery by ProxyQueryInterface.
+     *
+     * @param array<string, mixed> $hints
+     */
+    public static function create(ProxyQuery $proxyQuery, array $hints = []): Paginator
+    {
+        $queryBuilder = $proxyQuery->getQueryBuilder();
+
+        $identifierFieldNames = $queryBuilder
+            ->getEntityManager()
+            ->getClassMetadata(current($queryBuilder->getRootEntities()))
+            ->getIdentifierFieldNames();
+
+        $hasSingleIdentifierName = 1 === \count($identifierFieldNames);
+        $hasJoins = \count($queryBuilder->getDQLPart('join')) > 0;
+
+        $query = $proxyQuery->getDoctrineQuery();
+
+        if (!$hasJoins) {
+            $query->setHint(CountWalker::HINT_DISTINCT, false);
+        }
+
+        foreach ($hints as $name => $value) {
+            $query->setHint($name, $value);
+        }
+
+        // Paginator with fetchJoinCollection doesn't work with composite primary keys
+        // https://github.com/doctrine/orm/issues/2910
+        // To stay safe fetch join only when we have single primary key and joins
+        $paginator = new Paginator($query, $hasSingleIdentifierName && $hasJoins);
+
+        $hasHavingPart = null !== $queryBuilder->getDQLPart('having');
```

Comment:
I've change this part since [`having` part is `null` by default](https://github.com/doctrine/orm/blob/535bc92dc8280732cdbe1d6ed0a4034a6103996c/lib/Doctrine/ORM/QueryBuilder.php#L84) and [when it's filled it's a `Expr\Andx`](https://github.com/doctrine/orm/blob/535bc92dc8280732cdbe1d6ed0a4034a6103996c/lib/Doctrine/ORM/QueryBuilder.php#L1236-L1249).

Reply:
When calling `resetDQLPart` the having part is reset to `[]`.

Reply:
not really, see https://github.com/doctrine/orm/blob/535bc92dc8280732cdbe1d6ed0a4034a6103996c/lib/Doctrine/ORM/QueryBuilder.php#L1492-L1498

it depends on the current value

Reply:
Yes, so if you add an having part, it will become an array. And then when you reset it, since it's an array it will be reset to an empty array.

Reply:
I don't think so. The `having` part is never an array internally, or?

Reply:
I thought it was, but might be ok.